### PR TITLE
AB Tests: Allow testing for locales other than English or for another specific locale

### PR DESCRIPTION
**Pull request by yoavf**
Current abtests settings only allow for testing on "en" users, or all users using the `allowAnyLocale` option.

This PR introduces the ability to target "not-en" as well as any single specific locale by replacing `allowAnyLocale` with `targetLocales` which defaults to `en` but can also take `any`,`not-en`, or a specific locale slug as a value.

### CODE REVIEW

File: client/lib/abtest/index.js [hunk 1]
```diff
@@ -87,6 +88,13 @@ ABTest.prototype.init = function( name ) {
 		throw new Error( 'A default variation is specified for ' + name + ' but it is not part of the variations' );
 	}
 
+	const languageSlugs = map( config( 'languages' ), 'langSlug' );
+	const localeTargets = [ 'not-en', 'any' ].concat( languageSlugs );
```
Comment (hewsut) [inline client/lib/abtest/index.js]: What about `allowedLocaleTargets` here?

File: client/lib/abtest/index.js [hunk 2]
```diff
@@ -87,6 +88,13 @@ ABTest.prototype.init = function( name ) {
...
+
+	if ( testConfig.localeTargets && ! includes( localeTargets, testConfig.localeTargets ) ) {
+		throw new Error( 'localeTargets can by "any", "not-en" or any single locale slug.' );
```
Comment (hewsut) [inline client/lib/abtest/index.js]: Plural variable vs singular value confuses me here. I think I get why `localeTargets` is plural. In case of `any` or `not-en`? FWIW, I think you can get away with singular covering those (ie, `localeTarget`) which feels more consistent to me.

Comment (hewsut) [inline client/lib/abtest/index.js]: OTOH, from the other POV. After partially absorbing the rest of the code, it feels like the most flexible and intuitive thing would be two arrays: `localTargets` and `localeTargetExcludes` and then just diff values in `if` logic. Is that naive? Is there a restricting factor my drive-by view is missing that limits us to single values?

File: client/lib/abtest/index.js [hunk 3]
```diff
@@ -133,28 +154,34 @@ ABTest.prototype.isEligibleForAbTest = function() {
...
 
-	if ( ! this.allowAnyLocale ) {
-		if ( isUserSignedIn() && user.get().localeSlug !== 'en' ) {
-			debug( '%s: User has a non-English locale', this.experimentId );
-			return false;
+	if ( this.localeTarget ) {
+		const localeMatcher = new RegExp( '^' + this.localeTarget + '-?', 'i' );
+		let isTargetLocale = true;
+
+		if ( isUserSignedIn() && user.get().localeSlug !== this.localeTarget ) {
```
Comment (rralian) [inline client/lib/abtest/index.js]: I realize the prior code didn't do this either, but why not make use of `localeMatcher` for a more general `en-` match here (e.g., capture `en-gb` and `en-ca`, etc.)?

File: client/lib/abtest/index.js [hunk 4]
```diff
@@ -96,7 +104,20 @@ ABTest.prototype.init = function( name ) {
...
-	this.allowAnyLocale = testConfig.allowAnyLocale === true;
+
+	this.localeTarget = 'en';
+	this.localeTargetExclude = false;
+
+	if ( testConfig.localeTargets ) {
+		if ( testConfig.localeTargets === 'any' ) {
+			this.localeTarget = false;
+		} else if ( testConfig.localeTargets === 'not-en' ) {
+			this.localeTargetExclude = true;
```
Comment (hewsut) [inline client/lib/abtest/index.js]: So, if I'm following that `if`, the sole purpose of localeTargetExclude to exclude `en` right?  If yes, can we call it `excludeEnFromLocaleTarget` or`localeTargetExcludeEn` or something?

File: client/lib/abtest/index.js [hunk 5]
```diff
@@ -133,28 +154,34 @@ ABTest.prototype.isEligibleForAbTest = function() {
...
+			isTargetLocale = false;
+		}
+		if ( ! isUserSignedIn() && ! localeFromSession.match( localeMatcher ) ) {
+			debug( '%s: Logged-out user has the %s locale in session', this.experimentId, this.localeTarget );
+			isTargetLocale = false;
 		}
-		if ( ! isUserSignedIn() && ! localeFromSession.match( englishMatcher ) ) {
-			debug( '%s: Logged-out user has a non-English locale in session', this.experimentId );
+
+		if ( ( this.localeTargetExclude && isTargetLocale ) || ( ! this.localeTargetExclude && ! isTargetLocale ) ) {
```
Comment (hewsut) [inline client/lib/abtest/index.js]: Last q for tonight and this one maybe a dumb one from a fuzzy brain. What's the second half of that `or` doing in human speak?  Would removing the `or` and moving the following just below accomplish the same thing or am I looking at this wrong?

```
if ( ! isTargetLocale ) {
  return false;
}
```

File: client/lib/abtest/index.js (hunk 2 above)
Comment (yoavf) [inline client/lib/abtest/index.js]: Not naive, but I'm just trying to keep it simple for anyone who's going to use this, and try not to have to add code to handle human errors there.

**Comment by yoavf** [comment]
@hewsut thanks for the feedback - I simplifed the logic to just have an array that defaults to `['en']`. We can always populate this with all the locales except EN, but I don't see any actual use case for that :)

**Comment by yoavf** [comment]
It'd be nice to be able to target countries, not just locales.

File: client/lib/abtest/index.js [hunk 6]
```diff
@@ -87,6 +88,16 @@ ABTest.prototype.init = function( name ) {
 		throw new Error( 'A default variation is specified for ' + name + ' but it is not part of the variations' );
 	}
 
+	const languageSlugs = map( config( 'languages' ), 'langSlug' );
+
+	if (
+		testConfig.localeTargets &&
+		! every( testConfig.localeTargets, ( target ) => {
+			return languageSlugs.indexOf( target ) !== -1;
+		} ) ) {
```
Comment (hewsut) [inline client/lib/abtest/index.js]: Bear with as I muse/ramble. So, this throws an error if any `localTargets` are invalid (not in `languageSlugs`), correct? My first thought is that this `if` is certainly-not-impossible-but-a-little-cryptic to passerbys.  What about wrapping the test conditions in a simple function for readablity?  eg: 

`if ( this.testHasInvalidLocales( testConfig.localeTargets ) )`

File: client/lib/abtest/index.js [hunk 7]
```diff
@@ -87,6 +88,16 @@ ABTest.prototype.init = function( name ) {
...
+		throw new Error( 'localeTargets array contains invalid language slugs.' );
```
Comment (hewsut) [inline client/lib/abtest/index.js]: So, this forces a hard crash, ya?  Which is good in this case--because we want the A/B test setter to know there's a problem before the test gets committed?

File: client/lib/abtest/index.js [hunk 8]
```diff
@@ -96,7 +107,17 @@ ABTest.prototype.init = function( name ) {
...
+
+	this.localeTargets = [ 'en' ];
+
+	if ( 'undefined' !== typeof testConfig.localeTargets ) {
+		if ( testConfig.localeTargets === false ) {
+			this.localeTargets = false;
+		} else {
+			this.localeTargets = testConfig.localeTargets;
+		}
+	}
```
Comment (hewsut) [inline client/lib/abtest/index.js]: How do you feel about swapping in the string `'any'` for the behaviors you're triggering right now with `false`?  

Two reasons I ask:
1. Setting `testConfig.localeTargets = 'any'` feels relatively intuitive.
2. Then you can do something much simpler here like:
`this.localeTargets = testConfig.localeTargets || [ 'en' ];`

File: client/lib/abtest/index.js [hunk 9]
```diff
@@ -133,28 +154,29 @@ ABTest.prototype.isEligibleForAbTest = function() {
...
 
 	if ( ! store.enabled ) {
 		debug( '%s: Local storage is not enabled', this.experimentId );
 		return false;
 	}
 
-	if ( ! this.allowAnyLocale ) {
-		if ( isUserSignedIn() && user.get().localeSlug !== 'en' ) {
-			debug( '%s: User has a non-English locale', this.experimentId );
+	if ( this.localeTargets ) {
```
Comment (hewsut) [inline client/lib/abtest/index.js]: If you do switch to `any` you might have to instead `isArray` here or test explicitly `!== 'any'` or something.

File: client/lib/abtest/index.js (hunk 7 above)
Comment (yoavf) [inline client/lib/abtest/index.js]: Yep.

File: client/lib/abtest/index.js (hunk 8 above)
Comment (yoavf) [inline client/lib/abtest/index.js]: I'm trying to keep it simple - taking into account the current tests which all only run against "en". So no param = en, explicit false = all locales (not very likely that we'll ever do this imo), and array = one or more specific values.

File: client/lib/abtest/index.js (hunk 9 above)
Comment (yoavf) [inline client/lib/abtest/index.js]: Another reason why I'd prefer to keep it false rather than use "any"

File: client/lib/abtest/index.js (hunk 8 above)
Comment (hewsut) [inline client/lib/abtest/index.js]: FWIW, IMO, `localeTargets: false` is not simple from a devex POV. It's not the end of the world. I do think it adds a bit of cognitive overload that seems avoidable.

**Comment by hewsut** [comment]
If you've tested the code and feel good about it, I'd say 🚢

**Comment by yoavf** [comment]
@hewsut thought about it some more and went with your suggestions. Can you take another look?

**Comment by hewsut** [comment]
I'm still not crazy about `this.localeTargets = false` because it doesn't tell me the causal [dev] observer any sort of semantic "story" like `this.localeTargets = [ 'fr', 'de']` and `this.localeTargets = 'any'` do. I have to go read the README or trace over to some `if`s and stuff to have any clue as to why we're setting it false. 

But ya know we're really talking personal preference here to a fair degree I think. You have my blessing on this PR. Merge at your discretion. 👍

File: client/lib/abtest/index.js [hunk 10]
```diff
@@ -88,14 +91,15 @@ ABTest.prototype.init = function( name ) {
...
-		testConfig.localeTargets &&
-		! every( testConfig.localeTargets, ( target ) => {
-			return languageSlugs.indexOf( target ) !== -1;
-		} ) ) {
-		throw new Error( 'localeTargets array contains invalid language slugs.' );
+	this.localeTargets = [ 'en' ];
+	if ( testConfig.localeTargets ) {
+		if ( 'any' === testConfig.localeTargets ) {
+			this.localeTargets = false;
+		} else if ( isArray( testConfig.localeTargets ) && every( testConfig.localeTargets, langSlugIsValid ) ) {
```
Comment (hewsut) [inline client/lib/abtest/index.js]: Just to note: `every( testConfig.localeTargets, langSlugIsValid )` -- I do like that much more. 👍
